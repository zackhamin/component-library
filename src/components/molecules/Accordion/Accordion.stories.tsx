import { Story } from '@storybook/react';
import { Accordion, AccordionProps } from './Accordion';

export default {
	title: 'molecules/Accordion',
	component: Accordion,
};

const content = (
	<>
		<h1>Accordion content</h1>
		<p> Paragraph here</p>
	</>
);

const Template: Story<AccordionProps> = (args) => (
	<Accordion {...args} content={content} />
);

const defaultProps = {
	title: 'standard title as a string',
	subtitle: '',
	content: 'content can be components',
	expanded: false,
	disabled: false,
	testId: 'accordion',
	id: '',
	class: '',
	style: {},
} as AccordionProps;

export const Default = Template.bind({});
Default.args = {
	...defaultProps,
};

export const Disabled = Template.bind({});
Disabled.args = {
	...defaultProps,
	disabled: true,
};
export const Expanded = Template.bind({});
Expanded.args = {
	...defaultProps,
	expanded: true,
};

export const Subtitle = Template.bind({});
Subtitle.args = {
	...defaultProps,
	title: 'Title',
	subtitle: ' - Subtitle',
};
