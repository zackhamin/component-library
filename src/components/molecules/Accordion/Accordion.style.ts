import styled from 'styled-components';
import { baseStyles, colors, disabledStyles } from '../../../styles';

export const StyledAccordion = styled.div`
	${baseStyles};
	display: flex;
	flex-direction: column;
	overflow: hidden;
`;

export const StyledAccordionTitle = styled.div<{
	isExpanded?: boolean;
	disabled?: boolean;
}>`
	display: flex;
	align-items: center;
	justify-content: space-between;
	background-color: ${colors.transparent};
	padding: 0.4rem 1rem;
	border: 0;
	border-bottom: 1px solid ${colors.lightGrey};
	cursor: pointer;
	z-index: 1;
	margin: 0.125rem;
	h2 {
		font-size: 2.4rem;
		color: ${colors.offBlack};
	}
	h2 span {
		color: ${colors.darkGrey};
	}

	${({ disabled }) => disabled && ` ${disabledStyles}`}
`;

export const StyledAccordionArrow = styled.div<{ isExpanded?: boolean }>`
	width: 14px;
	height: 8px;
	margin-left: 0.5rem;
	filter: ${colors.emerald};
	transition: transform 0.5s ease-in-out;

	${({ isExpanded }) => isExpanded && `transform: rotate(180deg);`}
`;

export const StyledAccordionContent = styled.div<{ isExpanded?: boolean }>`
	height: 0;
	overflow: auto;
	visibility: hidden;
	padding: 0 1rem;

	-ms-overflow-style: none;
	scrollbar-width: none;

	&::-webkit-scrollbar {
		display: none;
	}
	${({ isExpanded }) =>
		isExpanded &&
		`border-bottom: 1px solid ${colors.lightGrey};
      padding: 0.4rem 1rem 2rem;
      height: 100%;  
	  display:block;
	  visibility: visible;
      `}
`;
