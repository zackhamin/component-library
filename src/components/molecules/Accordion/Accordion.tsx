import { Chevron } from 'Atoms/Icons';
import { isEventValid } from '../../../utils/utils';
import React, { CSSProperties, ReactNode, useRef, useState } from 'react';
import {
	StyledAccordion,
	StyledAccordionArrow,
	StyledAccordionContent,
	StyledAccordionTitle,
} from './Accordion.style';

export interface AccordionProps {
	id?: string;
	className?: string;
	style?: CSSProperties;
	title: string | ReactNode;
	content?: string | ReactNode;
	disabled?: boolean;
	expanded?: boolean;
	subtitle?: string | ReactNode;
	testId?: string;
}

export const Accordion: React.FC<AccordionProps> = ({
	title,
	content,
	disabled,
	expanded,
	subtitle,
	testId,
	...props
}) => {
	if (expanded != true) {
		expanded = false;
	}
	const [isExpanded, setIsExpanded] = useState<boolean>(expanded);
	const accordionRef = useRef<HTMLDivElement>(null);

	const toggleIsExpanded = (event: any) =>
		isEventValid(event) &&
		!disabled &&
		setIsExpanded((prevIsExpanded) => !prevIsExpanded);

	return (
		<StyledAccordion {...props} ref={accordionRef}>
			<StyledAccordionTitle
				role="button"
				tabIndex={disabled ? -1 : 0}
				aria-disabled={disabled}
				aria-expanded={isExpanded}
				onClick={toggleIsExpanded}
				onKeyDown={toggleIsExpanded}
				disabled={disabled}
				isExpanded={isExpanded}
				data-testid={testId}
			>
				<h2>
					{title}
					<span>{subtitle}</span>
				</h2>
				<StyledAccordionArrow isExpanded={isExpanded}>
					<Chevron fontSize="1.4rem" data-testid="CHEVRON" />
				</StyledAccordionArrow>
			</StyledAccordionTitle>

			<StyledAccordionContent isExpanded={isExpanded}>
				{content}
			</StyledAccordionContent>
		</StyledAccordion>
	);
};
