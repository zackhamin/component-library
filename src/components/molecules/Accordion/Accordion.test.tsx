import { Accordion } from './Accordion';
import React from 'react';
import { fireEvent, render } from '@testing-library/react';

describe('Accordion', () => {
	const defaultProps = {
		title: 'Accordion title',
		content: 'Accordion content',
	};

	test('renders without errors and matches snapshot', () => {
		const {
			container: { firstChild },
		} = render(<Accordion {...defaultProps} />);

		expect(firstChild).toMatchSnapshot();
	});

	test('should display title and content', () => {
		const { getByText } = render(<Accordion {...defaultProps} />);
		getByText(/title/i);
	});

	test('should show content on click', () => {
		const { getByText } = render(<Accordion {...defaultProps} />);
		const accordionTitle = getByText(/title/i);
		const accordionContent = getByText(/content/i);
		fireEvent.click(accordionTitle);

		expect(accordionContent).toBeVisible();
	});

	test('should set isExpanded to false if click outside of accordion', () => {
		const { getByText, getByTestId } = render(<Accordion {...defaultProps} />);
		const accordion = getByText(/title/i) as HTMLDivElement;
		fireEvent.click(accordion);
		const chevron = getByTestId('CHEVRON') as HTMLDivElement;
		fireEvent.mouseUp(document);
		const style = window.getComputedStyle(chevron);
		expect(style.transform).toBe('');
	});

	describe('Expanded Accordion', () => {
		test('should display title and content', () => {
			const { getByText } = render(
				<Accordion {...defaultProps} expanded={true} />
			);
			const accordionTitle = getByText(/title/i);
			const accordionContent = getByText(/content/i);
			expect(accordionTitle).toBeVisible();
			expect(accordionContent).toBeVisible();
		});
	});

	describe('Disabled Accordion', () => {
		test('should display title', () => {
			const { getByText } = render(
				<Accordion {...defaultProps} disabled={true} />
			);
			getByText(/title/i);
		});

		test('should be disabled', () => {
			const { getByRole } = render(<Accordion {...defaultProps} disabled />);
			expect(getByRole('button')).toHaveAttribute('disabled');
		});

		test('should not handle click events', () => {
			const { getByText } = render(
				<Accordion {...defaultProps} disabled={true} />
			);

			const accordionTitle = getByText(/title/i);
			const accordionContent = getByText(/content/i);
			fireEvent.click(accordionTitle);

			const titleStyle = window.getComputedStyle(accordionTitle);
			const contentStyle = window.getComputedStyle(accordionContent);
			expect(titleStyle.pointerEvents).toBe('');
			expect(contentStyle.visibility).toBe('hidden');
		});
	});
});
