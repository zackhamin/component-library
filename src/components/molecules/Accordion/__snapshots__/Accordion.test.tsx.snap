// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders without errors and matches snapshot 1`] = `
.c0 {
  font-family: Helvetica,Arial,sans-serif;
  font-size: 1rem;
  line-height: 1.2rem;
  margin: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
}

.c0:focus:not(:focus-visible) {
  outline: none;
}

.c0:focus-visible {
  outline: 0.125rem solid #212529;
}

.c0 h1 {
  color: #333;
  font-size: 3.2rem;
  line-height: 1.75;
  margin: 0 0 1rem;
}

.c0 h2 {
  color: #00594c;
  font-size: 2.4rem;
  line-height: 1.75;
  margin: 0 0 1rem;
}

.c0 p {
  font-size: 1.6rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  background-color: Transparent;
  padding: 0.4rem 1rem;
  border: 0;
  border-bottom: 1px solid #EEEEEF;
  cursor: pointer;
  z-index: 1;
  margin: 0.125rem;
}

.c1 h2 {
  font-size: 2.4rem;
  color: #333333;
}

.c1 h2 span {
  color: #585858;
}

.c2 {
  width: 14px;
  height: 8px;
  margin-left: 0.5rem;
  -webkit-filter: #00594C;
  filter: #00594C;
  -webkit-transition: -webkit-transform 0.5s ease-in-out;
  -webkit-transition: transform 0.5s ease-in-out;
  transition: transform 0.5s ease-in-out;
}

.c3 {
  height: 0;
  overflow: auto;
  visibility: hidden;
  padding: 0 1rem;
  -ms-overflow-style: none;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
}

.c3::-webkit-scrollbar {
  display: none;
}

<div
  class="c0"
>
  <div
    aria-expanded="false"
    class="c1"
    role="button"
    tabindex="0"
  >
    <h2>
      Accordion title
      <span />
    </h2>
    <div
      class="c2"
    >
      <svg
        aria-hidden="true"
        data-testid="CHEVRON"
        focusable="false"
        font-size="1.4rem"
        height="1em"
        viewBox="0 0 14 8"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M.293.293a1 1 0 0 1 1.414 0L7 5.586 12.293.293a1 1 0 1 1 1.414 1.414l-6 6a1 1 0 0 1-1.414 0l-6-6a1 1 0 0 1 0-1.414Z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
  <div
    class="c3"
  >
    Accordion content
  </div>
</div>
`;
